presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DetermineJoinDistributionType.java

```diff
@@ -132,18 +169,39 @@ private static boolean isCrossJoin(JoinNode node)
             return node.getType() == INNER && node.getCriteria().isEmpty();
         }
 
-        private SemiJoinNode.DistributionType getTargetSemiJoinDistributionType(boolean isDeleteQuery)
+        private SemiJoinNode.DistributionType getTargetSemiJoinDistributionType(SemiJoinNode semiJoinNode, boolean isDeleteQuery)
         {
-            if (isDistributedJoinEnabled(session) && !isDeleteQuery) {
+            // For delete queries, the TableScan node that corresponds to the table being deleted must be collocated with the Delete node,
+            // so you can't do a distributed semi-join
+            if (isDeleteQuery) {
+                return SemiJoinNode.DistributionType.REPLICATED;
+            }
+
+            // join distribution type forced by session variable
+            if (getJoinDistributionType(session).equals(FeaturesConfig.JoinDistributionType.REPLICATED)) {
+                return SemiJoinNode.DistributionType.REPLICATED;
+            }
+            if (getJoinDistributionType(session).equals(FeaturesConfig.JoinDistributionType.PARTITIONED)) {
                 return SemiJoinNode.DistributionType.PARTITIONED;
             }
 
-            return SemiJoinNode.DistributionType.REPLICATED;
+            // choose based on stats
+            if (isSmall(semiJoinNode.getFilteringSource())) {
+                return SemiJoinNode.DistributionType.REPLICATED;
+            }
+            return SemiJoinNode.DistributionType.PARTITIONED;
         }
 
-        private static boolean isDistributedJoinEnabled(Session session)
+        private boolean isSmall(PlanNode node)
         {
-            return !getJoinDistributionType(session).equals(FeaturesConfig.JoinDistributionType.BROADCAST);
+            double smallSizeLimit = .1 * globalProperties.getMaxMemoryPerNode().toBytes();
+            Estimate dataSize = costCalculator.calculateCostForNode(session, node).getOutputSizeInBytes();
+            if (!dataSize.isValueUnknown()) {
+                return dataSize.getValue() < smallSizeLimit;
+            }
+
+            dataSize = costCalculator.calculateCostForNode(session, node).getOutputRowCount();
+            return !dataSize.isValueUnknown() && dataSize.getValue() * ROW_SIZE_ESTIMATE < smallSizeLimit;
```

Comment:
Leave it for now but IMO the cost calculator should do the estimation by itself if connector does not exose explicit stats. Then this section will not be necessary.